### DISCUSSION TITLE
One image missing and one additional at the end?

### DISCUSSION THREAD

**Discussion by ultraviolet007rainbow** [nftchef/art-engine#85] [Help + Q&A]
Hi,

I just generated 10K again starting with 1, and ended up with 10,001 as the final image.
I went through and found out that there was a missing image at 7630.
So I edited the JSON of 10,001 to become 7630 instead as it should have been and made sure all the filenames were changed as well (JSON and Image). Will this affect the _dna.json in any way? I did identify the correct DNA line in the _dna.json that corresponds to image 10,0001 which is now image 7630. But I do not see anything in the _dna.json that would directly link it other than the traits to a specific filename (JSON or image). So is there a way to regenerate the _dna.json from the JSON files now that I have fixed the problem manually?

In addition, I found an oddity looking at the _dna.json where one of my background images I used has a numerical filename, none of the other layer files have numerical names other than this one. So in the _dna.json each one of those that has that specific background layer has 0.0 at the head of the line, which clearly is wrong. So for some reason the 0.0 was set because of the numerical layer name, is my assumption. 

Is there a good way to fix this problem?

My only idea is to rename the layer to something else in every JSON file as well as the actual layer name of that specific background PNG.

And also replace all instances in the _dna.json file of that numerical layer name.

Once again will this break anything, can I regenerate the _dna.json from the altered JSON files?

Thank you for your great improvements and enhancements to the original code.

Regards.

**Comment by nftchef** [comment]
Hi! this is a very interesting and odd situation. I would be happy to look into this more if you are able to share your configuration and project on Discord. genkihagata#3074

I've never seen the generator skip a number before. It's possible that there was an error while writing the file and it was never saved, but I  would need more data to reproduce it.

As for the _dna.json, it's only used for utility scripts, so it may not be that high priority to modify. It is an ordered _Set_ (array) of all the file names used to construct each token. To modify it, you would need to find the correct index for 7630 and insert the string from the 10,001 position.

Each file name in  _dna gets a prepended number _map_, this is how the generator navigates the subfolders. `0.0` indicates the first folder and the first subfolder for backgrounds.

**Reply by ultraviolet007rainbow** [reply]
Okay this project uses no sub-folders within the root layer folders.
But I understand what you mean.
So 0.0 makes sense then and is not wrong.
It simply means the first folder and the first file in that folder because it is the first filename in the background folder as it has a numerical name as opposed to all the others not. I thought it was an error but clearly it is not, thank you.

As far skipping a file, I see someone else recently had the same issue, where the generator creates 1 extra file at the end. 
As in mine 1-10,001 except there are only 10,000 files generated, which means it skipped a file somewhere.

The way I identified this was to copy and paste the filenames into Excel spreadsheet and line up the filenames with the row numbers on the left margin. Then I could fairly quickly by hand identify where the missing file appears. Because as soon as the filenames stop aligning with the row numbers I know that the missing file came before that spot and then I just narrow it down until I could find the missing number.

So are you saying manually editing the JSON file for 10,001 and making it 7,360 instead should work fine?

Also, since I identified the correct line in the DNA JSON file for 10,001, these were generated out of order to mix it up as I did have 4 different configurations. 4500 4500 500 500 = 10000.

How do I find the index in the array for 10,001 and change it to 7,360?

If I were to upload anything for you to check out you just need the config file?
I can't send all my metadata.

Thank you.

**Reply by ultraviolet007rainbow** [reply]
If I regenerate the metadata how does that work, it gets regenerated from the DNA JSON?
Can I regenerate the DNA JSON instead from all the JSON Metadata for the actual images?

**Reply by nftchef** [reply]
_metadata.json does not have any effect on the final collection, but like _dna.json, it's used in some util.
if you manually change the output individual JSON files, yes, you can regenerate the _metadata.js file using the `regenerateMetadata.js` script.

**Reply by ultraviolet007rainbow** [reply]
Okay the main thing is provenance.js
I did generate the hashes so I can generate the final provenance.
So is the DNA file used in that process?
What utils need the DNA file?
thx

**Reply by nftchef** [reply]
provenance relies on _metadata.json to read each items `imageHash`

_dna.json is not used in provenance, it is used in `regenerate` and `replaceAll`

**Reply by ultraviolet007rainbow** [reply]
thank you for your help with this :)

**Reply by ultraviolet007rainbow** [reply]
Okay just need to make sure because I didn't discuss this yet.
I was planning on using the util that inserts random ultra-rares into the collection.
Is this something I need the _dna.json for?
What is replaceAll? I am not seeing, I am only seeing replace. 
Oh wait you mean rebuildAll, yes? 
thx

**Reply by nftchef** [reply]
yes, sorry, `rebuildAll`
The replace util for ultra rares does not use _dna.json or modify it. any inserted rares should be done last and if you rebuild, they will be overwritten by what's stored in _dna.json. Perhaps there is room for improvement there or at least a note in the readme.

**Reply by vergara111** [reply]
indeed, it skips one number in the middle and creates extra one at the end
